File: ocaml/planet-time/lib/interplanet_time.ml
(* interplanet_time.ml — Public API for the Interplanetary Time Library
   Story 18.18 — OCaml planet-time library
   Mirrors the planet-time.js / C / Python APIs. *)

(* ── Re-exported types ───────────────────────────────────────────────────── *)

type planet_time = Time_calc.planet_time = {
  body                      : int;
  jd                        : float;
  sol                       : float;
  local_time_sec            : float;
  day_length_sec            : float;
  light_travel_from_earth_sec : float;
}

(* ── Body names ──────────────────────────────────────────────────────────── *)

(** [body_name body] returns the display name for body index 0..8. *)
let body_name (body : int) : string =
  if body >= 0 && body < Array.length Constants.body_names
  then Constants.body_names.(body)
  else "Unknown"

(* ── Julian Day ──────────────────────────────────────────────────────────── *)

(** [julian_day ~year ~month ~day ~hour ~minute ~second] converts a UTC
    calendar date/time to Julian Day number. *)
let julian_day = Orbital.julian_day

(* ── Mean longitude ──────────────────────────────────────────────────────── *)

(** [mean_longitude ~body ~jd] returns the mean longitude of a body
    in degrees for the given Julian Day. *)
let mean_longitude = Orbital.mean_longitude

(* ── True anomaly ────────────────────────────────────────────────────────── *)

(** [true_anomaly ~mean_anomaly ~eccentricity] solves Kepler's equation
    using Newton-Raphson (50 iterations) and returns the true anomaly
    in radians. *)
let true_anomaly = Orbital.true_anomaly

(* ── Ecliptic longitude ──────────────────────────────────────────────────── *)

(** [ecliptic_longitude ~body ~jd] returns the heliocentric ecliptic
    longitude of a body in degrees. *)
let ecliptic_longitude = Orbital.ecliptic_longitude

(* ── Heliocentric position ───────────────────────────────────────────────── *)

(** [heliocentric_pos ~body ~jd] returns the heliocentric (x, y, z)
    position in AU (ecliptic plane; z = 0). *)
let heliocentric_pos = Orbital.heliocentric_pos

(* ── Light travel time ───────────────────────────────────────────────────── *)

(** [light_travel_time ~body1 ~body2 ~jd] returns one-way light travel
    time in seconds between two bodies at the given Julian Day. *)
let light_travel_time = Orbital.light_travel_time

(* ── Solar day length ────────────────────────────────────────────────────── *)

(** [solar_day_seconds ~body] returns the solar day length in seconds
    for the given body. *)
let solar_day_seconds = Time_calc.solar_day_seconds

(* ── Local solar time ────────────────────────────────────────────────────── *)

(** [local_solar_time ~body ~jd ~longitude] returns seconds since
    midnight for the local solar time at the given surface longitude
    (degrees, positive = east). *)
let local_solar_time = Time_calc.local_solar_time

(* ── Sol number ──────────────────────────────────────────────────────────── *)

(** [sol_number ~body ~jd] returns the fractional sol/day number since
    the planet's epoch. *)
let sol_number = Time_calc.sol_number

(* ── Planet time ─────────────────────────────────────────────────────────── *)

(** [planet_time ~body ~unix_ms] computes the planet_time record for the
    given body at the given Unix millisecond timestamp (float). *)
let planet_time = Time_calc.planet_time

File: lua/planet-time/src/interplanet_time.lua
-- interplanet_time.lua — Public facade for the Interplanetary Time Lua library
-- Story 18.19
--
-- Usage:
--   local IPT = require("src.interplanet_time")
--   local result = IPT.planet_time(IPT.MARS, unix_ms)
--   local lt = IPT.light_travel_time(IPT.EARTH, IPT.MARS, unix_ms)

local C       = require("src.constants")
local Orbital = require("src.orbital")
local TimeCalc = require("src.time_calc")

local M = {}

M.VERSION = C.VERSION

-- ── Planet index constants ────────────────────────────────────────────────────

M.MERCURY = C.MERCURY
M.VENUS   = C.VENUS
M.EARTH   = C.EARTH
M.MARS    = C.MARS
M.JUPITER = C.JUPITER
M.SATURN  = C.SATURN
M.URANUS  = C.URANUS
M.NEPTUNE = C.NEPTUNE
M.MOON    = C.MOON

-- ── Re-export constants module ────────────────────────────────────────────────

M.PLANETS      = C.PLANETS
M.ORB_ELEMS    = C.ORB_ELEMS
M.PLANET_INDEX = C.PLANET_INDEX
M.AU_KM        = C.AU_KM
M.C_KMS        = C.C_KMS
M.AU_SECONDS   = C.AU_SECONDS
M.J2000_MS     = C.J2000_MS
M.J2000_JD     = C.J2000_JD
M.MARS_EPOCH_MS = C.MARS_EPOCH_MS
M.MARS_SOL_MS  = C.MARS_SOL_MS

-- ── Re-export orbital functions ───────────────────────────────────────────────

--- Julian Day from calendar components.
-- @param year, month, day, hour, minute, second
-- @return JD number
M.julian_day = Orbital.julian_day

--- Mean longitude of a body (degrees).
-- @param body_idx  Planet index
-- @param utc_ms   UTC milliseconds
M.mean_longitude = Orbital.mean_longitude

--- True anomaly from mean anomaly and eccentricity (degrees).
-- @param mean_anomaly_deg  degrees
-- @param eccentricity      dimensionless
M.true_anomaly = Orbital.true_anomaly

--- Ecliptic longitude of a body (degrees, 0..360).
-- @param body_idx  Planet index
-- @param utc_ms   UTC milliseconds
M.ecliptic_longitude = Orbital.ecliptic_longitude

--- Heliocentric (x, y, z) position in AU (ecliptic plane; z=0).
-- @param body_idx  Planet index
-- @param utc_ms   UTC milliseconds
-- @return x, y, z (AU)
M.heliocentric_pos = Orbital.heliocentric_pos

--- Heliocentric distance (r) of a body in AU.
-- @param body_idx  Planet index
-- @param utc_ms   UTC milliseconds
M.heliocentric_r = Orbital.heliocentric_r

--- One-way light travel time between two bodies (seconds).
-- @param body1_idx, body2_idx  Planet indices
-- @param utc_ms               UTC milliseconds
M.light_travel_time = Orbital.light_travel_time

-- ── Re-export time calculation functions ─────────────────────────────────────

--- Solar day length in seconds.
-- @param body_idx  Planet index
M.solar_day_seconds = TimeCalc.solar_day_seconds

--- Local solar time on a planet (seconds since midnight).
-- @param body_idx      Planet index
-- @param utc_ms        UTC milliseconds
-- @param longitude_deg Longitude in degrees (east positive)
M.local_solar_time = TimeCalc.local_solar_time

--- Sol number (total solar days since planet epoch).
-- @param body_idx  Planet index
-- @param utc_ms   UTC milliseconds
M.sol_number = TimeCalc.sol_number

--- Full planet-time record for a body at a UTC instant.
-- @param body_idx  Planet index
-- @param unix_ms   UTC milliseconds
-- @return table with all time fields
M.planet_time = TimeCalc.planet_time

return M
